packages/stores/src/impls/hashmap.rs

```diff
@@ -0,0 +1,289 @@
+use std::{
+    borrow::Borrow,
+    collections::HashMap,
+    hash::{BuildHasher, Hash},
+};
+
+use crate::store::Store;
+use dioxus_signals::{MappedMutSignal, ReadableExt, Writable};
+
+impl<
+        W: Writable<Target = HashMap<K, V, St>> + Copy + 'static,
+        K: 'static,
+        V: 'static,
+        St: 'static,
+    > Store<HashMap<K, V, St>, W>
+{
+    /// Get the length of the HashMap. This method will track the store shallowly and only cause
+    /// re-runs when items are added or removed from the map, not when existing values are modified.
+    ///
+    /// # Example
+    /// ```rust, no_run
+    /// use dioxus_stores::*;
+    /// use dioxus::prelude::*;
+    /// use std::collections::HashMap;
+    /// let mut store = use_store(|| HashMap::new());
+    /// assert_eq!(store.len(), 0);
+    /// store.insert(0, "value".to_string());
+    /// assert_eq!(store.len(), 1);
+    /// ```
+    pub fn len(self) -> usize {
+        self.selector().track_shallow();
+        self.selector().peek().len()
+    }
+
+    /// Check if the HashMap is empty. This method will track the store shallowly and only cause
+    /// re-runs when items are added or removed from the map, not when existing values are modified.
+    ///
+    /// # Example
+    /// ```rust, no_run
+    /// use dioxus_stores::*;
+    /// use dioxus::prelude::*;
+    /// use std::collections::HashMap;
+    /// let mut store = use_store(|| HashMap::new());
+    /// assert!(store.is_empty());
+    /// store.insert(0, "value".to_string());
+    /// assert!(!store.is_empty());
+    /// ```
+    pub fn is_empty(self) -> bool {
+        self.selector().track_shallow();
+        self.selector().peek().is_empty()
+    }
+
+    /// Iterate over the current entries in the HashMap, returning a tuple of the key and a store for the value. This method
+    /// will track the store shallowly and only cause re-runs when items are added or removed from the map, not when existing
+    /// values are modified.
+    ///
+    /// # Example
+    ///
+    /// ```rust, no_run
+    /// use dioxus_stores::*;
+    /// use dioxus::prelude::*;
+    /// use std::collections::HashMap;
+    /// let mut store = use_store(|| HashMap::new());
+    /// store.insert(0, "value1".to_string());
+    /// store.insert(1, "value2".to_string());
+    /// for (key, value_store) in store.iter() {
+    ///     println!("{}: {}", key, value_store.read());
+    /// }
+    /// ```
+    pub fn iter(
```

Comment:
for the iter methods, might be nice to add a "VecIter" iterator type to enable all the double-ended, exact-size, etc extensions

Reply:
If the only thing we want out of the custom type is the extra bounds, we could just make the return bounds more strict. Then we don't need to manually forward the implementations for all 4 traits for all of the iterators we expose